### PR TITLE
New entity editing

### PR DESCRIPTION
**Pull request by LordSputnik**
### Problem
The current entity editing forms in BookBrainz are overly complex. They don't flow well, use tabs although they aren't really needed, and it isn't really clear which information is required and which is optional.

### Solution
This PR introduces a new entity editor form for the creator entity, as a prototype that can be developed further based on feedback and extended to other entities prior to the next release. The prototype drops tabs, displays only a minimum of inputs by default and is based on last year's GCI UI concept drawings and concepts drawn up by myself and Leftmost 

### Areas of Impact
The client code for the creator editor has been replaced with this new code, and an additional function has been added to the server code processing the creator editor data to transform the new form data to the shape of the old form data. Style and syntax issues left over from #105 are also addressed.

### CODE REVIEW

File: src/client/entity-editor/common/error-text.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+import {Alert} from 'react-bootstrap';
+import React from 'react';
+import {connect} from 'react-redux';
+
+function ErrorText({
```
Comment (leftmostcat) [inline src/client/entity-editor/common/error-text.js]: `ErrorToast`?

File: src/client/entity-editor/common/language-field.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import VirtualizedSelect from 'react-virtualized-select';
+
+function LanguageField({
+	...props
+}) {
+	return (
+		<Input
+			label="Language"
+		>
+			<VirtualizedSelect {...props}/>
```
Comment (leftmostcat) [inline src/client/entity-editor/common/language-field.js]: It means more boilerplate, but how would you feel about explicitly specifying these props? It makes it a little difficult at this level to see what behavior will be affected at the `VirtualizedSelect` level.

File: src/client/entity-editor/common/name-field.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ */
+function NameField({
+	empty,
+	error,
+	onChange,
+	...rest
+}) {
+	const label = (
+		<ValidationLabel
+			empty={empty}
```
Comment (leftmostcat) [inline src/client/entity-editor/common/name-field.js]: Unclear what the purpose of this is from its name. Since it's a bool, I assume `isEmpty` is the intent.

File: src/client/entity-editor/common/name-field.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+			error={error}
```
Comment (leftmostcat) [inline src/client/entity-editor/common/name-field.js]: Since this is also a bool, `hasError`?

File: src/client/entity-editor/common/sort-name-field.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ * string.
+ *
+ * @param {String} name the input string to strip
+ * @returns {String} the string with dots removed
+ */
+function stripDot(name) {
+	return name.replace(/\./g, '');
+}
+
+function makeSortName(name) {
```
Comment (leftmostcat) [inline src/client/entity-editor/common/sort-name-field.js]: We'll want to give this code a more thorough review at some point. It may make the most sense to make it a specifiable parameter, since different entity types (and even different form field values) will require different behaviors.

File: src/client/entity-editor/common/sort-name-field.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	let lastName = words.splice(INDEX_BEFORE_END);
+	if (suffixWords.length > 0) {
+		lastName += ` ${suffixWords.join(' ')}`;
+	}
+
+	return `${lastName}, ${words.join(' ')}`;
+}
+
+/**
+ * Presentational component. Renders the sort name field for the alias section
```
Comment (leftmostcat) [inline src/client/entity-editor/common/sort-name-field.js]: I'm liking these descriptive components. I am in favor of adding them to the components that are missing them as well. 👍

File: src/client/entity-editor/common/submit-button.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+function submit(url, data) {
+	return request.post(url).send(data)
+		.promise()
+		.then((response) => {
+			if (!response.body) {
+				window.location.replace('/login');
+			}
+
+			const redirectUrl = `/creator/${response.body.bbid}`;
```
Comment (leftmostcat) [inline src/client/entity-editor/common/submit-button.js]: Since this is a generic entity form, we'll need to parameterize this.

File: src/client/entity-editor/common/submit-button.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+			{errorElement}
+		</div>
+	);
+}
+SubmitButton.displayName = 'SubmitButton';
+SubmitButton.propTypes = {
+	error: React.PropTypes.string
+};
+
+function submit(url, data) {
```
Comment (leftmostcat) [inline src/client/entity-editor/common/submit-button.js]: A lot of this behavior seems like it will be shared by all of our forms. The basic pattern of "check authentication, log in if lacking, continue". Seems like factoring this out would be good. We may also want to modalize login. All this is outside the scope of this PR, though.

File: src/client/entity-editor/creator-data.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+
+import AliasEditor from './alias-editor/alias-editor';
+import CreatorData from './creator-data/creator-data';
+import IdentifierEditor from './identifier-editor/identifier-editor';
+import {Panel} from 'react-bootstrap';
+import React from 'react';
+import SharedData from './shared-data/shared-data';
+import {connect} from 'react-redux';
+
```
Comment (leftmostcat) [inline src/client/entity-editor/creator-data.js]: #nit: Excess whitespace.

File: src/client/entity-editor/creator-data.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+let Wrapper = ({
```
Comment (leftmostcat) [inline src/client/entity-editor/creator-data.js]: Not sure I like the name `Wrapper`. If that's all this is, I wonder if it should be split into multiple components or subsumed by another.

File: src/client/entity-editor/creator-data/begin-field-container.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+import DateField from './date-field';
+import _debounce from 'lodash.debounce';
+import {connect} from 'react-redux';
+import {updateBeginDate} from '../actions';
+
+const KEYSTROKE_DEBOUNCE_TIME = 250;
+
+function isPartialDateValid(value) {
```
Comment (leftmostcat) [inline src/client/entity-editor/creator-data/begin-field-container.js]: This seems like functionality that belongs as part of the date field component.

File: src/client/entity-editor/creator-data/creator-data.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (leftmostcat) [inline src/client/entity-editor/creator-data/creator-data.js]: How does this file differ from `src/client/entity-editor/creator-data.js`? The naming is a little problematic.

File: src/client/entity-editor/shared-data/shared-data.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import SortNameField from './sort-name-field-container';
+
+function SharedData({
+	disambiguationVisible,
+	languageOptions
+}) {
+	return (
+		<div>
+			<h2>
+				What is the Creator called?
```
Comment (leftmostcat) [inline src/client/entity-editor/shared-data/shared-data.js]: Mention of creators in a generic form. Also, how do you feel about adding little hint popups with this sort of text next to a basic "Name" field?

File: src/server/app.js [hunk 14]
```diff
@@ -135,7 +135,8 @@ app.use((err, req, res, next) => { // eslint-disable-line no-unused-vars
 
 const debug = require('debug')('bbsite');
 
-app.set('port', process.env.PORT || 9099);
+const DEFAULT_PORT = 9099;
```
Comment (leftmostcat) [inline src/server/app.js]: Make it a config option?

File: src/server/helpers/achievement.js [hunk 15]
```diff
@@ -18,6 +18,8 @@
 
 'use strict';
 
+/* eslint prefer-spread: 1, prefer-reflect: 1, no-magic-numbers: 0 */
```
Comment (leftmostcat) [inline src/server/helpers/achievement.js]: Is there a particular reason these are specified here? Are they intended to be temporary pending style cleanup? If so, a `XXX` comment would be helpful.

File: src/server/helpers/achievement.js [hunk 16]
```diff
@@ -52,16 +58,16 @@ function awardUnlock(UnlockType, awardAttribs) {
...
-			else {
+			if (award === null) {
 				unlockPromise = new UnlockType(awardAttribs)
 					.save(null, {method: 'insert'})
 					.then((unlock) =>
 						unlock.toJSON()
 					);
 			}
+			else {
+				unlockPromise = Promise.resolve('already unlocked');
```
Comment (leftmostcat) [inline src/server/helpers/achievement.js]: Why resolve with a user-invisible string?

File: src/server/helpers/search.js [hunk 17]
```diff
@@ -132,6 +132,7 @@ search.indexEntity = (entity) =>
 search.refreshIndex = () =>
 	_client.indices.refresh({index: _index});
 
+/* eslint camelcase: 0, no-magic-numbers: 1 */
```
Comment (leftmostcat) [inline src/server/helpers/search.js]: Same question about eslint exceptions as above.

File: src/server/routes/entity/creator.js [hunk 18]
```diff
@@ -139,16 +139,54 @@ const additionalCreatorProps = [
...
 
-router.post('/create/handler', auth.isAuthenticatedForHandler, (req, res) =>
-	entityRoutes.createEntity(
-		req, res, 'Creator', _.pick(req.body, additionalCreatorProps)
-	)
-);
 
-router.post('/:bbid/edit/handler', auth.isAuthenticatedForHandler, (req, res) =>
-	entityRoutes.editEntity(
+function transformNewForm(data) {
```
Comment (leftmostcat) [inline src/server/routes/entity/creator.js]: Is the idea here to make this a temporary shim before we move to better data backend stuff? If so, please comment to that effect.

File: src/client/entity-editor/alias-editor/reducer.js [hunk 19]
```diff
@@ -16,6 +16,10 @@
  * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
 
+import {
+	ADD_ALIAS, REMOVE_ALIAS, UPDATE_ALIAS_LANGUAGE,
+	UPDATE_ALIAS_NAME, UPDATE_ALIAS_PRIMARY, UPDATE_ALIAS_SORT_NAME
+} from './actions';
 import Immutable from 'immutable';
 
 const DEFAULT_ALIAS = Immutable.Map({
```
Comment (leftmostcat) [inline src/client/entity-editor/alias-editor/reducer.js]: The naming of this confused me below. Instead of `DEFAULT`, how about `EMPTY`?

File: src/client/entity-editor/alias-editor/alias-editor.js [hunk 20]
```diff
@@ -16,16 +16,18 @@
...
+import AliasRowContainer from './alias-row-container';
 import React from 'react';
 import {connect} from 'react-redux';
-import {hideAliasEditor} from '../actions';
+
 
 const AliasEditor = ({
+	aliases,
 	languageOptions,
+	onAddButtonClick,
```
Comment (leftmostcat) [inline src/client/entity-editor/alias-editor/alias-editor.js]: Would it make sense to name this as `onAddAlias`? We only care about the button click internally. Externally, we only care that some action has been performed to add an alias.

**Comment by LordSputnik** [comment]
The importance of keeping ItemRow and ItemRowContainer separate is to ensure there's a separation of concerns between the layout (ItemRow) and the logic and state (ItemRowContainer). It may be possible to merge the container into the editor, but I'd need to have a look. It's also useful because it may be possible to make ItemRow the same component in both the alias and identifier editors in the future.

**Comment by leftmostcat** [comment]
I don't see a strong motivation for separating presentation and logic in this case, particularly given that the logic is a simple set of `mapDispatchToProps` and `mapStateToProps` call. I still think this is unnecessary indirection, and I would resist merging the presentation components of two distinct data types; it would complicate future changes. Aliases and identifiers aren't the related concepts; the sets are.

File: src/client/entity-editor/actions.js [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+	return {
+		type: ADD_ALIAS,
+		id: nextAliasId++
+	};
+}
+
+export function removeAlias(index) {
+	return {
+		type: REMOVE_ALIAS,
+		index
```
Comment (LordSputnik) [inline src/client/entity-editor/actions.js]: Nope, once an index is assigned, it stays with the associated alias.
